contrib/src/main/java/com/datatorrent/contrib/nifi/AbstractNiFiInputOperator.java

```
@@ -0,0 +1,162 @@
+ */
+package com.datatorrent.contrib.nifi;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.nifi.remote.Transaction;
+import org.apache.nifi.remote.TransferDirection;
+import org.apache.nifi.remote.client.SiteToSiteClient;
+import org.apache.nifi.remote.protocol.DataPacket;
+
+import com.datatorrent.api.Context;
+import com.datatorrent.api.InputOperator;
+
+/**
+ * This is the base implementation of a NiFi input operator.&nbsp;
+ * Subclasses should implement the methods which convert NiFi DataPackets to tuples and emit them.
+ * <p>
+ * Ports:<br>
+ * <b>Input</b>: No input port<br>
+ * <b>Output</b>: Can have any number of output ports<br>
+ * <br>
+ * Properties:<br>
+ * None<br>
+ * <br>
+ * Compile time checks:<br>
+ * Classes derived from this have to implement the abstract methods emitTuples(List<T> tuples)&nbsp;
+ * and createTuple(DataPacket dp)<br>
+ * <br>
+ * Run time checks:<br>
+ * None<br>
+ * <br>
+ * Benchmarks:<br>
+ * TBD<br>
+ * </p>
+ *
+ * @displayName Abstract NiFi Input
+ * @category Messaging
+ * @tags input operator
+ * @since 3.3.0
+ */
+
+public abstract class AbstractNiFiInputOperator<T> implements InputOperator
+{
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(AbstractNiFiInputOperator.class);
+
+  private SiteToSiteClient client;
+  private final SiteToSiteClient.Builder siteToSiteBuilder;
+
+  /**
+   * @param siteToSiteBuilder the builder for a NiFi SiteToSiteClient
+   */
+  public AbstractNiFiInputOperator(final SiteToSiteClient.Builder siteToSiteBuilder)
+  {
+    this.siteToSiteBuilder = siteToSiteBuilder;
+  }
+
+  @Override
+  public void setup(Context.OperatorContext context)
+  {
+    this.client = siteToSiteBuilder.build();
+  }
+
+  @Override
+  public void teardown()
+  {
+    try {
+      client.close();
+    } catch (IOException e) {
+      LOGGER.error(e.getMessage(), e);
```

don't swallow exception